src/mds/Server.cc

```diff
@@ -5432,26 +5432,45 @@ int Server::parse_layout_vxattr(string name, string value, const OSDMap& osdmap,
 				file_layout_t *layout, bool validate)
 {
   dout(20) << "parse_layout_vxattr name " << name << " value '" << value << "'" << dendl;
+  auto parse_pool = [&](std::string& pool) -> int64_t {
+    try {
+      return boost::lexical_cast<int64_t>(pool);
+    } catch (boost::bad_lexical_cast const&) {
+      int64_t pool_id = osdmap.lookup_pg_pool_name(pool);
+      if (pool_id < 0) {
+	dout(10) << " unknown pool " << pool << dendl;
+	return -CEPHFS_ENOENT;
+      }
+      return pool_id;
+    }
+  };
   try {
     if (name == "layout") {
-      string::iterator begin = value.begin();
-      string::iterator end = value.end();
-      keys_and_values<string::iterator> p;    // create instance of parser
-      std::map<string, string> m;             // map to receive results
-      if (!qi::parse(begin, end, p, m)) {     // returns true if successful
-	return -CEPHFS_EINVAL;
-      }
-      string left(begin, end);
-      dout(10) << " parsed " << m << " left '" << left << "'" << dendl;
-      if (begin != end)
+      JSONParser json_parser;
```

Comment:
Would this break the old way of setting layout?

Reply:
@vshankar yes; this would break the old way of setting layouts

On another note, @batrick suggested to keep things the old way just in case the new client happened to talk to an old server. So, code in `src/client/Client.cc` tests for a return value of `-ENOSYS` to detect an old server. But practically this doesn't seem to work since the Nautilus as well as the Pacific servers crash when trying to collect op latency stats for the new `getvxattr` op. So the `-ENOSYS` doesn't reach the client to make use of the old method of setting layouts.

@batrick also suggested to change only the entire layout get/set mechanism to JSON and keep the old way for setting individual fields.

Reply:
> @vshankar yes; this would break the old way of setting layouts
> 
> On another note, @batrick suggested to keep things the old way just in case the new client happened to talk to an old server. So, code in `src/client/Client.cc` tests for a return value of `-ENOSYS` to detect an old server. But practically this doesn't seem to work since the Nautilus as well as the Pacific servers crash when trying to collect op latency stats for the new `getvxattr` op. So the `-ENOSYS` doesn't reach the client to make use of the old method of setting layouts.

The two cases we have to worry about is::

- new client talking to an old server: the client sends the data in json format and the server errors out. The client retries with the old way of setting layout xattrs? Note that the client does not get an ENOTSUPP since its operation is supported by the server, but the xattr value format has changed.
- old client talking to a new server: we still need to preserve the old way if setting layouts

> 
> @batrick also suggested to change only the entire layout get/set mechanism to JSON and keep the old way for setting individual fields.

What about setting the layout xattr at once? Shouldn't we preserve that too?

Reply:
please review earlier comments; this concern has been addressed